include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,400 @@
+        else    // fail if samples have invalid paths
+            return sample_type::createInvalid();    // should check if sample direction is invalid
+    }
+    template<typename C=bool_constant<IsBSDF> >
+    enable_if_t<C::value && IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(anisocache_type) cache)
+    {
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const scalar_type NdotV = localV.z;
+
+        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, NdotV);
+        fresnel::OrientedEtaRcps<monochrome_type> rcpEta = _f.getOrientedEtaRcps();
+
+        const vector3_type upperHemisphereV = ieee754::flipSignIfRHSNegative<vector3_type>(localV, hlsl::promote<vector3_type>(NdotV));
+        const vector3_type localH = ndf.generateH(upperHemisphereV, u.xy);
+        const scalar_type VdotH = hlsl::dot(localV, localH);
+
+        NBL_IF_CONSTEXPR(!ndf_type::GuaranteedVNDF)
+        {
+            // allow for rejection sampling, theoretically NdotV=0 or VdotH=0 is valid, but leads to 0 value contribution anyway
+            if ((IsBSDF ? (NdotV*VdotH) : VdotH) <= scalar_type(0.0))
+                return sample_type::createInvalid();
+            assert(!hlsl::isnan(NdotV*VdotH));
+        }
+        else
+        {
+            assert(NdotV*VdotH >= scalar_type(0.0));
+        }
+        const scalar_type reflectance = _f(hlsl::abs(VdotH))[0];
+
+        scalar_type rcpChoiceProb;
+        scalar_type z = u.z;
+        bool transmitted = math::partitionRandVariable(reflectance, z, rcpChoiceProb);
+        const scalar_type LdotH = hlsl::mix(VdotH, ieee754::copySign(hlsl::sqrt(rcpEta.value2[0]*VdotH*VdotH + scalar_type(1.0) - rcpEta.value2[0]), -VdotH), transmitted);
+
+        // fail if samples have invalid paths
+        const scalar_type NdotL = hlsl::mix(scalar_type(2.0) * VdotH * localH.z - NdotV,
+                                    localH.z * (VdotH * rcpEta.value[0] + LdotH) - NdotV * rcpEta.value[0], transmitted);
+        // VNDF sampling guarantees that `VdotH` has same sign as `NdotV`
+        // and `transmitted` controls the sign of `LdotH` relative to `VdotH` by construction (reflect -> same sign, or refract -> opposite sign)
+        if (ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(NdotV, NdotL) != transmitted)
+            return sample_type::createInvalid();    // should check if sample direction is invalid
+
+        cache = anisocache_type::createPartial(VdotH, LdotH, localH.z, transmitted, rcpEta);
+        assert(cache.isValid(fresnel::OrientedEtas<vector_type>::create(scalar_type(1.0), hlsl::promote<vector_type>(_f.getRefractionOrientedEta()))));
+
+        ray_dir_info_type V = interaction.getV();
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+        assert(hlsl::abs(hlsl::length(H) - scalar_type(1.0)) < scalar_type(1e-4));
```

Comment:
btw assertion for orthonormality of matrix is just checking the dot products between all pairs of rows are 1.0

You could have a general utility function `bool dotIsUnity(A,B);` then this becomes
```cpp
// tangent frame orthonormality
assert(dotIsUnity(fromTangent[0],fromTangent[1]));
assert(dotIsUnity(fromTangent[1],fromTangent[2]));
assert(dotIsUnity(fromTangent[2],fromTangent[0]));
// NDF sampling produced a unit length direction
assert(dotIsUnity(localH,localH));
```
you don't want to assert stuff on the output H